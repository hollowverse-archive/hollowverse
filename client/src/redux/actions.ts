//
// REDUX ACTION CREATORS
//
//
// This file contains:
// * Redux action creators
// * Redux action types
//
import {push} from 'react-router-redux'
import {INotablePersonSchema, IUserSchema} from '../../../typings/dataSchema'
import {HvError} from '../../../typings/typeDefinitions'
import {stringEnum} from '../utils/utils'
import {IAlgoliaSearchResults} from '../vendor/algolia'

// Custom type definition of a Redux Action
export interface IAction<PayloadType> { type: string, payload: PayloadType}

// The following are all the actions that can be triggered from within the Hollowverse application
export const actions = {
  navigateToSearch: (payload: string) => {
    return push({
      pathname: '/',
      search: payload,
    })
  },

  // Regular action Types
  setSearchInputValue: (payload: string) => ({type: 'setSearchInputValue', payload}),
  setIsSearchPending: (payload: boolean) => ({type: 'setIsSearchPending', payload}),
  setIsLoginPending: (payload: boolean) => ({type: 'setIsLoginPending', payload}),
  setIsLogoutPending: (payload: boolean) => ({type: 'setIsLogoutPending', payload}),
  setSearchError: (payload: string) => ({type: 'setSearchError', payload}),
  setSearchResults: (payload: IAlgoliaSearchResults | undefined) => ({type: 'setSearchResults', payload}),
  setLoginStatus: (payload: facebookSdk.LoginStatus | undefined) => ({type: 'setLoginStatus', payload}),
  setError: (payload: HvError) => ({type: 'setError', payload}),
  setIsNavMenuOpen: (payload: boolean) => ({type: 'setIsNavMenuOpen', payload}),
  setLastSearchTerm: (payload: string) => ({type: 'setLastSearchTerm', payload}),
  setCreateProfileUrlInputValue: (payload: string) => ({type: 'setCreateProfileUrlInputValue', payload}),
  setNotablePerson: (payload: INotablePersonSchema | undefined) => ({type: 'setNotablePerson', payload}),
  setUserData: (payload: IUserSchema | undefined) => ({type: 'setUserData', payload}),
  toggleWarning: (payload: boolean) => ({type: 'toggleWarning', payload}),

  // Redux Sagas triggers
  requestSearchResults: (payload: string) => ({type: 'requestSearchResults', payload}),
  requestLogin: () => ({type: 'requestLogin'}),
  requestLogout: () => ({type: 'requestLogout'}),
  requestUpdateLoginStatus: () => ({type: 'requestUpdateLoginStatus'}),
  requestNotablePerson: (payload: string) => ({type: 'requestNotablePerson', payload}),
  requestUserData: (payload: string) => ({ type: 'requestUserData', payload}),
}

// `ActionTypes` can have any string value that corresponds with a key in the `actions` object above
// so `navigateToSearch` | `setSearchInputValue` | `setIsLogoutPending` | etc...are all valid.
export type ActionTypes = keyof typeof actions

// Instead of maintaining a separate list of Redux Action Types, we can use the keys of our `actions`
// object above as our action types. That's what `stringEnum` does.
//
// Its type is being asserted as you can see so that when you try to access a key that doesn't exist
// on the `actions` object, you'll get a compile error.
export const ActionTypes = stringEnum(Object.keys(actions)) as {[name in ActionTypes]: ActionTypes}
