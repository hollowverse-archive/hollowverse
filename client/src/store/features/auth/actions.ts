import { createActionCreator } from 'store/helpers';

export const requestLogin = createActionCreator('requestLogin');
export const setLoginStatus = createActionCreator('setLoginStatus');
export const requestUpdateLoginStatus = createActionCreator(
  'requestUpdateLoginStatus',
);
export const setIsLoginPending = createActionCreator('setIsLoginPending');
export const setIsLogoutPending = createActionCreator('setIsLogoutPending');
export const requestLogout = createActionCreator('requestLogout');
