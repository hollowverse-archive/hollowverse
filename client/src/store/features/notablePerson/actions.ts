import { createActionCreator } from 'store/helpers';

export const requestNotablePerson = createActionCreator('requestNotablePerson');
