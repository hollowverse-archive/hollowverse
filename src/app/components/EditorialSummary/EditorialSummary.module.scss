@import 'variables.scss';

.article-text {
  padding: $spacing-20px;
}

.source-list {
  padding: 0 $spacing-30px;
  li {
    margin-left: $spacing-10px;
    margin-top: $spacing-10px;
  }
}

.back-link {
  border-radius: 50%;
  margin-left: $spacing-5px;
  padding: $spacing-1px;
  display: inline-block;
  &:hover {
    text-decoration: none;
    background-color: $color-blue-opacity;
  }
}
