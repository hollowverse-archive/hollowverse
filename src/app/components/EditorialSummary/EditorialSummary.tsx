import React from 'react';
import formatDate from 'date-fns/format';

import classes from './EditorialSummary.module.scss';
import { prettifyUrl } from 'helpers/prettifyUrl';
import { EditorialSummaryNodeType } from 'api/types';
import { Quote } from 'components/Quote/Quote';
import { Collapsable } from 'components/Collapsable/Collapsable';

type Node = {
  id: string;
  parentId: string | null;
  text: string | null;
  type: EditorialSummaryNodeType;
  sourceUrl: string | null;
  sourceTitle: string | null;
};

type Props = {
  id: string;
  author: string;
  lastUpdatedOn: string | null;
  nodes: Node[];
};

const findChildren = (node: Node, nodes: Node[]) => {
  return nodes.filter(child => child.parentId === node.id);
};

const isBlockNode = (node: Node) => {
  return [
    EditorialSummaryNodeType.heading,
    EditorialSummaryNodeType.paragraph,
    EditorialSummaryNodeType.quote,
  ].includes(node.type);
};

const isRootBlock = (node: Node) => !node.parentId;

type Source = {
  number: number;
  sourceId: string;
  nodeId: string;
  sourceUrl: string;
  sourceTitle: string | null;
};

type BlockProps = {
  node: Node;
  nodes: Node[];
  referencesMap: Map<Node, Source>;
  onSourceClick?: React.EventHandler<React.MouseEvent<HTMLAnchorElement>>;
};

const Block = (props: BlockProps): JSX.Element => {
  const { node, nodes, referencesMap, onSourceClick } = props;
  const children = findChildren(node, nodes).map(child => {
    if (isBlockNode(child)) {
      return <Block key={child.id} {...props} node={child} />;
    } else if (child.type === 'link' && child.sourceUrl) {
      return (
        <a
          key={child.id}
          title={child.sourceTitle || undefined}
          href={child.sourceUrl}
        >
          {child.text}
        </a>
      );
    }

    const ref = referencesMap.get(child);

    return (
      <span key={child.id}>
        {child.type === 'emphasis' ? <em>{child.text}</em> : child.text}
        {ref ? (
          <a id={ref.nodeId} onClick={onSourceClick} href={`#${ref.sourceId}`}>
            <sup>{ref.number}</sup>
          </a>
        ) : null}
      </span>
    );
  });

  if (node.type === 'quote') {
    return (
      <Quote key={node.id} size="large">
        {children}
      </Quote>
    );
  } else if (node.type === 'heading') {
    return <h3 key={node.id}>{children}</h3>;
  }

  return <p key={node.id}>{children}</p>;
};

/**
 * Finds and adds references to `map` by performing a depth-first search
 * for nodes with `sourceUrl`s.
 *
 * Why depth-first search?
 * ----------------------
 * Sources should follow the same order of their reference positions.
 * We traverse the nodes by the order they would appear on the screen and
 * collect the sources they are found.
 */
const findRefs = (nodes: Node[], map: Map<Node, Source>) => (node: Node) => {
  const { sourceUrl, sourceTitle, type } = node;
  if (sourceUrl && type !== 'link') {
    const number = map.size + 1;
    map.set(node, {
      number,
      sourceId: `source_${number}`,
      nodeId: `ref_${number}`,
      sourceUrl: sourceUrl,
      sourceTitle,
    });
  }

  findChildren(node, nodes).forEach(findRefs(nodes, map));
};

type State = {
  shouldShowSources: boolean;
};

export class EditorialSummary extends React.PureComponent<Props, State> {
  references = new Map<Node, Source>();

  state: State = {
    shouldShowSources: true,
  };

  constructor(props: Props, context: any) {
    super(props, context);

    const { nodes } = props;

    // Collect references in nodes to create the Sources section at the end
    // of the editorial summary.
    nodes.filter(isRootBlock).forEach(findRefs(nodes, this.references));
  }

  componentDidMount() {
    // The server markup should have the sources shown by default (see default `state` above)
    // so that users can click on a reference in the content to go the source.
    // even if they have disabled JavaScript in the browser.
    //
    // When JS is executed on the client, React calls `componentDidMount`
    // and sees that, `shouldShowSources` is now `false`,
    // so it hides them.
    this.setState({ shouldShowSources: false });
  }

  onSourceClick: BlockProps['onSourceClick'] = () => {
    this.setState({ shouldShowSources: true });
  };

  render() {
    const { nodes, author, lastUpdatedOn } = this.props;
    const { shouldShowSources } = this.state;

    const date = lastUpdatedOn ? new Date(lastUpdatedOn) : undefined;

    return (
      <div className={classes.root}>
        {nodes
          .filter(isRootBlock)
          .map(node => (
            <Block
              key={node.id}
              node={node}
              nodes={nodes}
              referencesMap={this.references}
              onSourceClick={this.onSourceClick}
            />
          ))}
        <hr />
        <Collapsable
          className={classes.sourceListContainer}
          isOpen={shouldShowSources}
          label={<h3>Sources</h3>}
        >
          <small>
            <ol className={classes.sourceList}>
              {Array.from(this.references.values()).map(ref => {
                const { nodeId, sourceId, sourceUrl, sourceTitle } = ref;
                const prettifiedUrl = prettifyUrl(sourceUrl);

                return (
                  <li key={nodeId} id={sourceId}>
                    <a href={sourceUrl}>{sourceTitle || prettifiedUrl}</a>
                    {sourceTitle ? ` ${prettifiedUrl}` : null}
                    <a
                      className={classes.backLink}
                      href={`#${nodeId}`}
                      role="button"
                      aria-label="Go back to reference"
                    >
                      â†©&#xFE0E;
                    </a>
                  </li>
                );
              })}
            </ol>
          </small>
        </Collapsable>
        <hr />
        <footer className={classes.footer}>
          <small>
            This article was written by {author}
            {date ? (
              <time dateTime={date.toISOString()}>
                {' '}
                and was last updated on {formatDate(date, 'MMMM D, YYYY')}
              </time>
            ) : null}.
          </small>
        </footer>
      </div>
    );
  }
}
