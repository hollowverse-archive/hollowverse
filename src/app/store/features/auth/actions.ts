import { createActionCreator } from 'store/helpers';

export const toggleAuthStatus = createActionCreator('TOGGLE_AUTH_STATUS');
export const facebookAuthResponseChanged = createActionCreator(
  'FACEBOOK_AUTH_TOKEN_CHANGED',
);
