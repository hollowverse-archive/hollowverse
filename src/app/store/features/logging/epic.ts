import { merge, fromEvent as observableFromEvent } from 'rxjs';

import {
  buffer,
  ignoreElements,
  tap,
  bufferCount,
  map,
  distinctUntilChanged,
  withLatestFrom,
  filter,
} from 'rxjs/operators';

import { Action, StoreState, LoggedAction } from 'store/types';

import { Epic } from 'redux-observable';

import {
  pageLoadSucceeded,
  pageLoadFailed,
  pageRedirected,
} from 'store/features/logging/actions';
import { LOCATION_CHANGE } from 'react-router-redux';
import { createPath } from 'history';
import { isEqual, pick } from 'lodash';

import { EpicDependencies } from 'store/createConfiguredStore';

const loggableActions: Array<Action['type']> = [
  'PAGE_LOAD_SUCCEEDED',
  'PAGE_LOAD_FAILED',
  'PAGE_REDIRECTED',
  'UNHANDLED_ERROR_THROWN',
  'SEARCH_RESULT_SELECTED',
  'TOGGLE_NIGHT_MODE',
];

const shouldActionBeLogged = (action: Action) =>
  loggableActions.includes(action.type);

const locationCompareKeys = ['search', 'pathname', 'hash'];

const mapPageLoadActions = ([setStatusCodeAction, locationChangeAction]: [
  Action<'SET_STATUS_CODE'>,
  Action<typeof LOCATION_CHANGE>
]) => {
  const path = createPath(locationChangeAction.payload);
  const statusCodePayload = setStatusCodeAction.payload;

  if (statusCodePayload.code === 301 || statusCodePayload.code === 302) {
    const to = statusCodePayload.redirectTo;

    return pageRedirected({
      from: path,
      to,
      statusCode: statusCodePayload.code,
    });
  } else if (statusCodePayload.code === 500) {
    return pageLoadFailed({ path, error: statusCodePayload.error });
  }

  return pageLoadSucceeded({ path });
};

const comparePageLoadActions = (
  x: [Action<'SET_STATUS_CODE'>, Action<typeof LOCATION_CHANGE>],
  y: typeof x,
) => {
  const [setStatusCodeActionX, locationChangeActionX] = x;
  const [setStatusCodeActionY, locationChangeActionY] = y;
  const setStatusCodeActionPayloadX = setStatusCodeActionX.payload;
  const setStatusCodeActionPayloadY = setStatusCodeActionY.payload;
  const locationChangePayloadX = locationChangeActionX.payload;
  const locationChangePayloadY = locationChangeActionY.payload;
  const areStatusCodeActionsEqual = isEqual(
    setStatusCodeActionPayloadX,
    setStatusCodeActionPayloadY,
  );

  if (
    locationChangePayloadX.pathname === '/search' &&
    locationChangePayloadY.pathname === '/search'
  ) {
    // When on search page, do not log a page visit every time
    // the query in the URL changes, only look at the pathname instead
    //
    // This is to avoid sending too many logs while the user is typing
    // the search query, as the URL will keep changing to reflect the
    // search term: /search?query=t, /search?query=to, /search?query=tom
    return areStatusCodeActionsEqual;
  }

  return (
    areStatusCodeActionsEqual &&
    isEqual(
      pick(locationChangePayloadX, locationCompareKeys),
      pick(locationChangePayloadY, locationCompareKeys),
    )
  );
};

export const loggingEpic: Epic<Action, StoreState, EpicDependencies> = (
  action$,
  _state$,
  { sendLogs, getSessionId, getUserAgent },
) => {
  const observePageLoad$ = action$.ofType('SET_STATUS_CODE').pipe(
    withLatestFrom(action$.ofType(LOCATION_CHANGE)),
    distinctUntilChanged<any>(comparePageLoadActions),
    map(mapPageLoadActions),
  );

  const createLoggableActionsObserver = () => {
    const loggableActions$ = action$.pipe(
      filter(shouldActionBeLogged),
      map(
        (action): LoggedAction => ({
          ...action,
          timestamp: new Date(),
        }),
      ),
    );

    const flushOnUnload$ = merge(
      // `pagehide` is for Safari
      observableFromEvent(window, 'pagehide'),
      observableFromEvent(window, 'unload'),
    );

    const logOnUnload$ = loggableActions$.pipe(buffer(flushOnUnload$));
    const logOnIdle$ = loggableActions$.pipe(bufferCount(10));

    return merge(logOnIdle$, logOnUnload$).pipe(
      filter(actions => actions.length > 0),
      tap(async actions => {
        await sendLogs({
          actions,
          sessionId: getSessionId(),
          userAgent: getUserAgent(),
        });
      }),
      ignoreElements(),
    );
  };

  return merge(observePageLoad$, createLoggableActionsObserver());
};
