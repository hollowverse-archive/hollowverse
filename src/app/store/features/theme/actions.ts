import { createActionCreator } from 'store/helpers';

export const toggleNightMode = createActionCreator('TOGGLE_NIGHT_MODE');
