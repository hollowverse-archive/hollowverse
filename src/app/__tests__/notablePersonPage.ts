import {
  ClientSideTestContext,
  createClientSideTestContext,
} from 'helpers/testHelpers';
import { pageLoadSucceeded } from 'store/features/logging/actions';
import {
  notablePersonWithEditorialSummaryQueryResponse,
  stubNotablePersonQueryResponse,
} from 'fixtures/notablePersonQuery';
import { EditorialSummary } from 'components/EditorialSummary/EditorialSummary';

describe('Notable Person page', () => {
  let context: ClientSideTestContext;

  describe('When notable person is found,', () => {
    beforeEach(async () => {
      context = await createClientSideTestContext({
        mockDataResponsesOverrides: {
          notablePersonQuery: stubNotablePersonQueryResponse,
        },
        createHistoryOptions: { initialEntries: ['/Tom_Hanks'] },
      });
    });

    it('has notable person name', () => {
      expect(context.wrapper).toIncludeText('Tom Hanks');
    });

    it('shows related people', () => {
      expect(context.wrapper).toIncludeText('Al Pacino');
    });

    describe('logs page load event', () => {
      beforeEach(done => {
        window.addEventListener('unload', () => {
          done();
        });

        window.dispatchEvent(new Event('unload'));
      });

      it('sends logs on page unload', () => {
        expect(context.dependencies.sendLogs).toHaveBeenLastCalledWith(
          expect.arrayContaining([
            pageLoadSucceeded(
              context.history.createHref(context.history.location),
            ),
          ]),
        );
      });
    });

    describe('sends analytics', () => {
      it('loads Google Analytics script', () => {
        expect(
          context.dependencies.getGoogleAnalyticsFunction,
        ).toHaveBeenCalled();
      });

      it('sets the account settings', async () => {
        const ga = await context.dependencies.getGoogleAnalyticsFunction();

        expect(ga).toHaveBeenCalledWith(
          'create',
          expect.stringMatching(/UA-[0-9]{9}-[0-9]{1,2}/g),
          expect.nonEmptyString(),
        );
      });

      it('sets the active page correctly', async () => {
        const ga = await context.dependencies.getGoogleAnalyticsFunction();

        expect(ga).toHaveBeenCalledWith(
          'set',
          'page',
          context.history.location.pathname,
        );
      });

      it('sends pageview event', async () => {
        const ga = await context.dependencies.getGoogleAnalyticsFunction();

        expect(ga).toHaveBeenLastCalledWith(
          'send',
          'pageview',
          context.history.createHref(context.history.location),
        );
      });
    });

    describe('if notable person does not have an editorial summary', () => {
      it('shows a call to comment', () => {
        expect(context.wrapper).toIncludeText('Share what you know');
      });
    });

    describe('if notable person has an editorial summary', () => {
      beforeEach(async () => {
        context = await createClientSideTestContext({
          mockDataResponsesOverrides: {
            notablePersonQuery: notablePersonWithEditorialSummaryQueryResponse,
          },
          createHistoryOptions: { initialEntries: ['/Tom_Hanks'] },
        });
      });

      it('shows editorial summary content', () => {
        const editorialSummary = context.wrapper.find(EditorialSummary);
        expect(editorialSummary).toBePresent();
        expect(editorialSummary).toMatchSnapshot();
      });
    });

    describe('if notable person has an image', () => {
      beforeEach(async () => {
        context = await createClientSideTestContext({
          mockDataResponsesOverrides: {
            notablePersonQuery: {
              ...stubNotablePersonQueryResponse,
              notablePerson: {
                ...stubNotablePersonQueryResponse.notablePerson!,
                mainPhoto: {
                  url:
                    'https://files.hollowverse.com/notable-people/Tom_Hanks.jpg',
                  colorPalette: null,
                  sourceUrl:
                    'https://commons.wikimedia.org/wiki/File:Tom_Hanks_2014.jpg',
                },
              },
            },
          },
          createHistoryOptions: { initialEntries: ['/Tom_Hanks'] },
        });
      });

      it('shows notable person image', () => {
        expect(
          context.wrapper.find(
            `img[src="https://files.hollowverse.com/notable-people/Tom_Hanks.jpg"]`,
          ),
        ).toBePresent();
      });

      it('page includes attribution link', () => {
        expect(
          context.wrapper.find(
            `a[href="https://commons.wikimedia.org/wiki/File:Tom_Hanks_2014.jpg"]`,
          ),
        ).toBePresent();
      });
    });
  });

  describe('When notable person is not found,', () => {
    beforeEach(async () => {
      context = await createClientSideTestContext({
        createHistoryOptions: { initialEntries: ['/Tom_Hanks'] },
        mockDataResponsesOverrides: {
          notablePersonQuery: {
            notablePerson: null,
          },
        },
      });
    });

    it('shows "Not Found"', () => {
      expect(context.wrapper).toIncludeText('Not Found');
    });
  });

  describe('On load failure', () => {
    beforeEach(async () => {
      context = await createClientSideTestContext({
        createHistoryOptions: { initialEntries: ['/Tom_Hanks'] },
        epicDependenciesOverrides: {
          getResponseForDataRequest: async payload => {
            if (payload.key === 'notablePersonQuery') {
              throw new TypeError();
            }

            return payload.load();
          },
        },
      });
    });

    it('offers to reload', () => {
      const linkButton = context.wrapper.findWhere(
        el => el.is('a') && Boolean(el.text().match(/reload/i)),
      );
      expect(linkButton).toBePresent();
      expect(linkButton.render().attr('href')).toMatch('/Tom_Hanks');
    });
  });

  describe('Links to other pages on the website', () => {
    beforeAll(async () => {
      context = await createClientSideTestContext({
        createHistoryOptions: { initialEntries: ['/Tom_Hanks'] },
      });
    });

    it('has a link to search page', () => {
      expect(context.wrapper.find('a[href="/search"]')).toBePresent();
    });
  });
});
