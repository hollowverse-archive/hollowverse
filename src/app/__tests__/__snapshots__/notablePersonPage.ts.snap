// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notable person page when notable person is found, if notable person has an editorial summary shows editorial summary content 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded MuiCard-root editorialSummary"
>
  <div
    class="articleText"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
    >
      <span>
        Tom Hanks was born in Concord, California.
      </span>
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
    >
      <span>
        His religious beliefs are about as diverse as the characters he has played. Hanks' childhood was a hodgepodge of divorce and step-parents, each with their own views. Hanks has identified with everything from Mormonism to his most recent and enduring faith–Greek Orthodox. In his own words:
      </span>
    </p>
    <blockquote
      class="root large"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
      >
        <span>
          The major religion I was exposed to in the first 10 years of my life was Catholicism. My stepmother became a Mormon. My aunt, whom I lived with for a long time, was a Nazarene, which is kind of ultra-super Methodist, and in high school, all my friends were Jews. For years I went to Wednesday-night Bible studies with my church group. So I had this peripatetic overview of various faiths, and the one thing I got from that was the intellectual pursuit involved. There was a lot of great stuff to think about.
          <a
            href="#source_1"
            id="ref_1"
          >
            <sup>
              1
            </sup>
          </a>
        </span>
      </p>
    </blockquote>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
    >
      <span>
        Hanks' second and current wife, Rita Wilson, was raised a member of the Greek Orthodox Church and Hanks has taken on her religion.
      </span>
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
    >
      <span>
        Hanks is said to be a deeply religious man and goes to church regularly.
        <a
          href="#source_2"
          id="ref_2"
        >
          <sup>
            2
          </sup>
        </a>
      </span>
      <span>
         Hanks has said that his greatest motivation for being a church-goer is to come in contact with the great unanswered questions mankind has always asked, questions that, for the most part, only religion attempts to answer.
        <a
          href="#source_3"
          id="ref_3"
        >
          <sup>
            3
          </sup>
        </a>
      </span>
    </p>
    <h2
      class="MuiTypography-root MuiTypography-title MuiTypography-gutterBottom"
    >
      <span>
        Politics is just like a box of chocolates…
      </span>
    </h2>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
    >
      <span>
        Politically, Hanks falls squarely in the Democrat/liberal camp, like the vast majority of his Hollywood counterparts. He's stuck up for quite a few liberal causes like gay marriage and the environment, and has a large collection of electric and hybrid cars.
      </span>
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
    >
      <span>
        He even slandered his old church–The Church of Jesus Christ of Latter Day Saints (Mormons) when they donated money to advance proposition 8 in California, effectively banning gay marriage.
        <a
          href="#source_4"
          id="ref_4"
        >
          <sup>
            4
          </sup>
        </a>
      </span>
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
    >
      <span>
        However, he quickly rescinded his statement but fought fire with fire, contributing his star power to raising over $44 million to defeat proposition 8–which passed anyway.
        <a
          href="#source_5"
          id="ref_5"
        >
          <sup>
            5
          </sup>
        </a>
      </span>
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
    >
      <span>
        Hanks backed Obama in 2008, posting a video on his Myspace page, explaining his position.
        <a
          href="#source_6"
          id="ref_6"
        >
          <sup>
            6
          </sup>
        </a>
      </span>
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
    >
      <span>
        In 2012, Hanks forcefully said he would back Obama once again and, despite widespread opinion to the opposite effect, said Obama had exceeded his expectations, saying:
      </span>
    </p>
    <blockquote
      class="root large"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
      >
        <span>
          If you would have told me a few years ago that 'don't ask, don't tell' would be repealed and about a billion jobs at General Motors and Chrysler would have been saved because the president was smart enough and strong enough and bold enough to do so, I would have said, 'Wow. That's a good president, I think I'll vote for him again.'
          <a
            href="#source_7"
            id="ref_7"
          >
            <sup>
              7
            </sup>
          </a>
        </span>
      </p>
    </blockquote>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
    >
      <span>
        He doesn't seem to acknowledge the billions upon billions of taxpayer dollars that were funneled to some of the country's richest men during that bailout, but hey, that's what the PR machine is all about.
      </span>
    </p>
    <footer>
      <small
        class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
      >
        This article was written by 
        Tom Kershaw
        <time
          datetime="2012-04-05T00:00:00.000Z"
        >
           
          and was last updated on 
          April 5, 2012
        </time>
        .
      </small>
    </footer>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation0 MuiExpansionPanel-root"
  >
    <div
      aria-expanded="false"
      class="MuiButtonBase-root MuiExpansionPanelSummary-root"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiExpansionPanelSummary-content"
      >
        <h3
          class="MuiTypography-root MuiTypography-subheading"
        >
          Sources
        </h3>
      </div>
      <div
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiExpansionPanelSummary-expandIcon"
        role="button"
        tabindex="-1"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div
      aria-hidden="true"
      class="MuiCollapse-container"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            class="MuiExpansionPanelDetails-root"
          >
            <small>
              <ol
                class="sourceList"
              >
                <li
                  id="source_1"
                >
                  <a
                    href="http://www.johnsanidopoulos.com/2009/05/orthodoxy-of-tom-hanks.html"
                  >
                    Mystagogy: The Orthodoxy of Tom Hanks
                  </a>
                   johnsanidopoulos.com
                  <a
                    aria-label="Go back to reference"
                    class="backLink"
                    href="#ref_1"
                    role="button"
                  >
                    ↩︎
                  </a>
                </li>
                <li
                  id="source_2"
                >
                  <a
                    href="http://www.beliefnet.com/Celebrity-Faith-Database/H/Tom-Hanks.aspx"
                  >
                    Tom Hanks
                  </a>
                   beliefnet.com
                  <a
                    aria-label="Go back to reference"
                    class="backLink"
                    href="#ref_2"
                    role="button"
                  >
                    ↩︎
                  </a>
                </li>
                <li
                  id="source_3"
                >
                  <a
                    href="http://www.johnsanidopoulos.com/2009/05/orthodoxy-of-tom-hanks.html"
                  >
                    Mystagogy: The Orthodoxy of Tom Hanks
                  </a>
                   johnsanidopoulos.com
                  <a
                    aria-label="Go back to reference"
                    class="backLink"
                    href="#ref_3"
                    role="button"
                  >
                    ↩︎
                  </a>
                </li>
                <li
                  id="source_4"
                >
                  <a
                    href="http://mormonmatters.org/2009/03/11/does-tom-hanks-hate-mormons/"
                  >
                    Does Tom Hanks Hate Mormons?
                  </a>
                   mormonmatters.org
                  <a
                    aria-label="Go back to reference"
                    class="backLink"
                    href="#ref_4"
                    role="button"
                  >
                    ↩︎
                  </a>
                </li>
                <li
                  id="source_5"
                >
                  <a
                    href="http://www.latimes.com/news/local/la-moneymap,0,2198220.htmlstory"
                  >
                    Proposition 8: Tracking the Money
                  </a>
                   latimes.com
                  <a
                    aria-label="Go back to reference"
                    class="backLink"
                    href="#ref_5"
                    role="button"
                  >
                    ↩︎
                  </a>
                </li>
                <li
                  id="source_6"
                >
                  <a
                    href="http://www.myspace.com/video/vid/33546997"
                  >
                    Beware: Celebrity Endorsement Video by Tom Hanks
                  </a>
                   myspace.com
                  <a
                    aria-label="Go back to reference"
                    class="backLink"
                    href="#ref_6"
                    role="button"
                  >
                    ↩︎
                  </a>
                </li>
                <li
                  id="source_7"
                >
                  <a
                    href="http://www.politico.com/blogs/click/0611/Tom_Hanks_Im_voting_for_Obama_again.html"
                  >
                    Tom Hanks: I'm Voting for Obama Again
                  </a>
                   politico.com
                  <a
                    aria-label="Go back to reference"
                    class="backLink"
                    href="#ref_7"
                    role="button"
                  >
                    ↩︎
                  </a>
                </li>
              </ol>
            </small>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
